src/sentry/static/sentry/app/components/grid.jsx

```diff
@@ -0,0 +1,217 @@
+    pt: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
+    flex: PropTypes.string,
+    alignItems: PropTypes.string,
+    justifyContent: PropTypes.string,
+    flexDirection: PropTypes.string,
+    flexWrap: PropTypes.string,
+  };
+
+  getSpacingRules = () => {
+    const spacingProps = [
+      ['m', ['margin']],
+      ['mx', ['margin-left', 'margin-right']],
+      ['my', ['margin-top', 'margin-bottom']],
+      ['ml', ['margin-left']],
+      ['mr', ['margin-right']],
+      ['mt', ['margin-top']],
+      ['mb', ['margin-bottom']],
+      ['p', ['padding']],
+      ['px', ['padding-left', 'padding-right']],
+      ['py', ['padding-top', 'padding-bottom']],
+      ['pl', ['padding-left']],
+      ['pr', ['margin-right']],
+      ['pt', ['padding-top']],
+      ['pb', ['padding-bottom']],
+    ];
+
+    const rules = [];
+
+    spacingProps.forEach(([prop, cssProperties]) => {
+      const val = this.getSpacingValue(this.props[prop]);
+      if (typeof val !== 'undefined') {
+        cssProperties.forEach(property => {
+          rules.push(`${property}:${val};`);
+        });
+      }
+    });
+
+    return rules;
+  };
+
+  getSpacingValue = prop => {
+    switch (typeof prop) {
+      case 'number':
+        return space(prop);
+      case 'string':
+        return prop;
+      case 'undefined':
+      default:
+        return undefined;
+    }
+  };
+
+  getWidthRules = () => {
+    const rules = [];
+    const type = typeof this.props.w;
+    if (type === 'number') {
+      rules.push(`width:${this.getWidthValue(this.props.w)};`);
+    } else if (Array.isArray(this.props.w) && this.props.w.length > 0) {
+      const firstVal = this.getWidthValue(this.props.w[0]);
+
+      rules.push(`width:${this.getWidthValue(firstVal)};`);
+
+      this.props.w.slice(1).forEach((wVal, idx) => {
+        rules.push(
+          `@media(min-width:${theme.breakpoints[idx]}){width:${this.getWidthValue(
+            wVal
+          )}};`
+        );
+      });
+    } else if (type === 'string') {
+      rules.push(`width:${this.props.w};`);
+    }
+
+    return rules;
+  };
+
+  getWidthValue = num => {
+    if (typeof num === 'number') {
+      const isPx = num >= 1;
+      return isPx ? `${num}px` : `${num * 100}%`;
+    } else {
+      return num;
+    }
+  };
+
+  getFlexRules = () => {
+    const rules = [];
+    const mapping = [
+      ['alignItems', 'align-items'],
+      ['justifyContent', 'justify-content'],
+      ['flexDirection', 'flex-direction'],
+      ['flexWrap', 'flex-wrap'],
+    ];
+
+    mapping.forEach((prop, cssProperty) => {
+      if (typeof this.props[prop] !== 'undefined') {
+        rules.push(`${cssProperty}:${this.props[prop]};`);
+      }
+    });
+    return rules;
+  };
+
+  getFlexChildRules = () => {
+    const rules = [];
+    if (typeof this.props.flex !== 'undefined') {
+      rules.push(`flex:${this.props.flex};`);
+    }
+
+    return rules;
+  };
+}
+
+export class Box extends Base {
```

Comment:
second crazy proposal: do we need two different components here? They seem to share the majority of their props. 

It seems like we could just infer a flex layout by looking for other flex properties, or just include a `flex` attribute.